contracts/voting/cwd-voting-cw721-staked/src/state.rs

```diff
@@ -1,43 +1,98 @@
 use cw721_controllers::NftClaims;
 use cw_controllers::Hooks;
-use cw_storage_plus::{Item, SnapshotItem, SnapshotMap, Strategy};
+use cw_storage_plus::{Item, Map, SnapshotItem, SnapshotMap, Strategy};
 use cw_utils::Duration;
-use indexmap::set::IndexSet;
+
+use crate::ContractError;
 
 #[cw_serde]
 pub struct Config {
     pub owner: Option<Addr>,
-    pub manager: Option<Addr>,
     pub nft_address: Addr,
     pub unstaking_duration: Option<Duration>,
 }
 
 pub const CONFIG: Item<Config> = Item::new("config");
 
-/// Maps addresses to the set of NFTs they have staked with this
-/// contract at a given height.
-///
-/// We use an IndexSet here to get linear time pagination queries.
-pub const STAKED_NFTS_PER_OWNER: SnapshotMap<Addr, IndexSet<String>> = SnapshotMap::new(
-    "staked_nfts_per_owner",
-    "staked_nfts_per_owner__checkpoints",
-    "staked_nfts_per_owner__changelog",
+/// The set of NFTs currently staked by each address. The existence of
+/// an `(address, token_id)` pair implies that `address` has staked
+/// `token_id`.
+pub const STAKED_NFTS_PER_OWNER: Map<(&Addr, &str), Empty> = Map::new("snpw");
+/// The number of NFTs staked by an address as a function of block
+/// height.
+pub const NFT_BALANCES: SnapshotMap<&Addr, Uint128> = SnapshotMap::new(
+    "nb",
+    "nb__checkpoints",
+    "nb__changelog",
     Strategy::EveryBlock,
 );
-
-/// The number of NFTs staked with this contract at a given height.
+/// The number of NFTs staked with this contract as a function of
+/// block height.
 pub const TOTAL_STAKED_NFTS: SnapshotItem<Uint128> = SnapshotItem::new(
-    "total_staked_nfts",
-    "total_staked_nfts__checkpoints",
-    "total_staked_nfts__changelog",
+    "tsn",
+    "tsn__checkpoints",
+    "tsn__changelog",
     Strategy::EveryBlock,
 );
 
 /// The maximum number of claims that may be outstanding.
 pub const MAX_CLAIMS: u64 = 100;
 pub const NFT_CLAIMS: NftClaims = NftClaims::new("nft_claims");
 
-// Hooks to contracts that will receive staking and unstaking messages
+// Hooks to contracts that will receive staking and unstaking
+// messages.
 pub const HOOKS: Hooks = Hooks::new("hooks");
+
+pub fn register_staked_nft(
+    storage: &mut dyn Storage,
+    height: u64,
+    staker: &Addr,
+    token_id: &String,
+) -> StdResult<()> {
+    let add_one = |prev: Option<Uint128>| -> StdResult<Uint128> {
+        prev.unwrap_or_default()
+            .checked_add(Uint128::new(1))
+            .map_err(StdError::overflow)
+    };
+
+    STAKED_NFTS_PER_OWNER.save(storage, (staker, token_id), &Empty::default())?;
+    NFT_BALANCES.update(storage, staker, height, add_one)?;
+    TOTAL_STAKED_NFTS
+        .update(storage, height, add_one)
+        .map(|_| ())
+}
+
+pub fn register_unstaked_nft(
+    storage: &mut dyn Storage,
+    height: u64,
+    staker: &Addr,
+    token_ids: &[String],
+) -> Result<(), ContractError> {
+    let subtractor = |amount: u128| {
```

Comment:
i love this

Reply:
aha yeah always exciting when you get to use a higher order function :)